packages/component-base/src/slot-observer.js

```diff
@@ -0,0 +1,56 @@
+/**
+ * @license
+ * Copyright (c) 2023 Vaadin Ltd.
+ * This program is available under Apache License Version 2.0, available at https://vaadin.com/license/
+ */
+
+/**
+ * A helper for observing slot changes.
+ */
+export class SlotObserver {
+  constructor(slot, callback) {
+    /** @type HTMLSlotElement */
+    this.slot = slot;
+
+    /** @type Function */
+    this.callback = callback;
+
+    /** @type {Node[]} */
+    this._storedNodes = [];
+
+    this._processNodes();
+
+    slot.addEventListener('slotchange', () => {
+      this._processNodes();
+    });
+  }
+
+  /**
+   * Run the observer callback synchronously.
+   */
+  flush() {
+    this._processNodes();
+  }
+
+  /** @private */
+  _processNodes() {
+    const currentNodes = this.slot.assignedNodes({ flatten: true });
```

Comment:
I wonder if using flatten is correct. Are we interested in nodes that are assigned to children slots?

Reply:
We generally use flattened nodes, see #4337. Also, this is what `FlattenedNodesObserver` does.

Reply:
Got it. Thanks!